lib/THC/generic/THCTensorCopy.cu

```diff
@@ -4,251 +4,44 @@
-      THCTensor_(copy)(state, self, (THCTensor*) src);   /* cast just removes compiler warning */ \
-    } else {                                                            \
-      THArgCheck(THCTensor_(nElement)(state, self) == THCuda##TYPECUDA##Tensor_nElement(state, src), 2, "size does not match");    \
-      THLongStorage *size = THCuda##TYPECUDA##Tensor_newSizeOf(state, src);   \
-      TH##TYPEC##Tensor *buffer1 = TH##TYPEC##Tensor_newWithSize(size, NULL); \
-      THTensor *buffer2  = THTensor_(newWithSize)(size, NULL);          \
-      TH##TYPEC##Tensor_copyCuda(state, buffer1, src);                  \
-      THTensor_(copy##TYPEC)(buffer2, buffer1);                         \
-      THCTensor_(copyCPU)(state, self, buffer2);                        \
-      THLongStorage_free(size);                                         \
-      TH##TYPEC##Tensor_free(buffer1);                                  \
-      THTensor_(free)(buffer2);                                         \
-    }                                                                   \
-  }
-#else
-#define THC_CUDA_TENSOR_IMPLEMENT_COPY(TYPEC,TYPECUDA)                  \
-  void THCTensor_(copyCuda##TYPEC)(THCState *state, THCTensor *self, struct THCuda##TYPECUDA##Tensor *src)  \
-  {                                                                     \
-    THArgCheck(THCTensor_(nElement)(state, self) == THCuda##TYPECUDA##Tensor_nElement(state, src), 2, "size does not match");    \
-    if (THCTypeIdx_(TYPEC) == THCTypeIdxFloat) {                        \
-      THCudaTensor *csrc = THCudaTensor_newContiguous(state, (THCudaTensor*) src); /* cast removes compiler error */             \
-      THCFloat2Half(state,                                              \
-          THCTensor_(data)(state, self),                                \
-          THCudaTensor_data(state, csrc),                               \
-          THCudaTensor_nElement(state, csrc));                          \
-      THCudaTensor_free(state, csrc);                                   \
-    } else {                                                            \
-      THLongStorage *size = THCuda##TYPECUDA##Tensor_newSizeOf(state, src); \
-      THCudaTensor *buffer = THCudaTensor_newWithSize(state, size, NULL); \
-      THCudaTensor_copyCuda##TYPEC(state, buffer, src);                 \
-      THCFloat2Half(state,                                              \
-          THCTensor_(data)(state, self),                                \
-          THCudaTensor_data(state, buffer),                             \
-          THCudaTensor_nElement(state, buffer));                        \
-      THCudaTensor_free(state, buffer);                                 \
-      THLongStorage_free(size);                                         \
-    }                                                                   \
+#define IMPLEMENT_THC_CUDA_TENSOR_COPY(TYPEC, TYPECUDA)                 \
```

Comment:
copy from, say, CudaTensorDouble to CudaTensorByte now happens exclusively on the GPU